File: frontend/lib/features/inspections/domain/inspection.dart
enum InspectionStatus { planned, inProgress, completed, cancelled }

extension InspectionStatusValue on InspectionStatus {
  String get apiValue => switch (this) {
    InspectionStatus.planned => 'planned',
    InspectionStatus.inProgress => 'in_progress',
    InspectionStatus.completed => 'completed',
    InspectionStatus.cancelled => 'cancelled',
  };
  String get label => switch (this) {
    InspectionStatus.planned => 'Planowana',
    InspectionStatus.inProgress => 'W toku',
    InspectionStatus.completed => 'Zakończona',
    InspectionStatus.cancelled => 'Anulowana',
  };
}

class Inspection {
  const Inspection({
    required this.id,
    this.projectId,
    this.projectName,
    required this.clientId,
    required this.clientName,
    required this.title,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.scheduledAt,
    this.startedAt,
    this.completedAt,
    this.notes,
    this.latitude,
    this.longitude,
    this.locationAccuracyM,
  });
  final int id;
  final int? projectId;
  final String? projectName;
  final int clientId;
  final String clientName;
  final String title;
  final InspectionStatus status;
  final DateTime? scheduledAt;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final String? notes;
  final double? latitude;
  final double? longitude;
  final double? locationAccuracyM;
  final DateTime createdAt;
  final DateTime updatedAt;
  String get location => latitude == null || longitude == null
      ? 'brak'
      : '${latitude!.toStringAsFixed(5)}, ${longitude!.toStringAsFixed(5)}';
  factory Inspection.fromJson(Map<String, dynamic> json) => Inspection(
    id: json['id'] as int,
    projectId: json['project_id'] as int?,
    projectName: json['project_name']?.toString(),
    clientId: json['client_id'] as int,
    clientName: json['client_name']?.toString() ?? '',
    title: json['title']?.toString() ?? '',
    status: InspectionStatus.values.firstWhere(
      (value) => value.apiValue == json['status'],
      orElse: () => InspectionStatus.planned,
    ),
    scheduledAt: DateTime.tryParse(json['scheduled_at']?.toString() ?? ''),
    startedAt: DateTime.tryParse(json['started_at']?.toString() ?? ''),
    completedAt: DateTime.tryParse(json['completed_at']?.toString() ?? ''),
    notes: json['notes']?.toString(),
    latitude: (json['latitude'] as num?)?.toDouble(),
    longitude: (json['longitude'] as num?)?.toDouble(),
    locationAccuracyM: (json['location_accuracy_m'] as num?)?.toDouble(),
    createdAt: DateTime.parse(json['created_at'].toString()),
    updatedAt: DateTime.parse(json['updated_at'].toString()),
  );
}

class InspectionPage {
  const InspectionPage({
    required this.items,
    required this.total,
    required this.skip,
    required this.limit,
  });
  final List<Inspection> items;
  final int total;
  final int skip;
  final int limit;
}

File: backend/app/schemas/inspection.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

InspectionStatus = Literal["planned", "in_progress", "completed", "cancelled"]


class InspectionBase(BaseModel):
    client_id: int = Field(gt=0)
    status: InspectionStatus = "planned"
    scheduled_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    notes: str | None = Field(default=None, max_length=10_000)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    location_accuracy_m: float | None = Field(default=None, ge=0)

    @field_validator("notes", mode="before")
    @classmethod
    def strip_strings(cls, value: object) -> object:
        if isinstance(value, str):
            value = value.strip()
            return value or None
        return value

    @model_validator(mode="after")
    def validate_timing_and_location(self):
        if (self.latitude is None) != (self.longitude is None):
            raise ValueError("latitude and longitude must be provided together")
        if (
            self.started_at is not None
            and self.completed_at is not None
            and self.completed_at < self.started_at
        ):
            raise ValueError("completed_at cannot precede started_at")
        if self.status != "completed" and self.completed_at is not None:
            raise ValueError("completed_at is only valid for completed inspections")
        return self


class InspectionCreate(InspectionBase):
    pass


class InspectionUpdate(BaseModel):
    client_id: int | None = Field(default=None, gt=0)
    status: InspectionStatus | None = None
    scheduled_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    notes: str | None = Field(default=None, max_length=10_000)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    location_accuracy_m: float | None = Field(default=None, ge=0)


class InspectionRead(InspectionBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    project_id: int | None
    project_name: str | None
    title: str
    client_name: str
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class InspectionPage(BaseModel):
    items: list[InspectionRead]
    total: int
    skip: int
    limit: int
